### PR TITLE
Free rebalance (roll) - initial implementation

### PR DESCRIPTION
**Pull request by talbaneth**

### CODE REVIEW

File: src/CharterManager.sol [hunk 1]
```diff
@@ -229,6 +242,28 @@ contract CharterManagerImp {
         VatLike(vat).move(urp, dst, rad);
     }
 
+    function roll(bytes32 srcIlk, bytes32 dstIlk, address src, address dst, int256 srcDart) external allowed(src) allowed(dst) {
```
Comment (hexonaut) [inline src/CharterManager.sol:245]: Gonza mentioned changing `srcDart` to a `uint256` in chat, and I think I agree with him. `int256` is a little more generalized, but it's practically a confusing parameter. Additionally even if we allow positive `srcDart` for a reverse transfer (I'm assuming this is the intent?) we probably want to have explicit permission set for rebalancing in either direction. If we allow positive `srcDart` then `rollable` permissions only need to be set in one direction and you get both.

**Review by hexonaut — COMMENTED** [review]
Looks correct. A point about the type of `srcDart`, and I think there should be more test coverage.

File: src/CharterManager.sol (hunk 1 above)
Comment (talbaneth) [reply src/CharterManager.sol]: Ok, changed `srcDart` to uint256.
The original intention was too keep it similar to `frob`, as now we have `dart`s with different types in different functions.
But it's probably ok.

**Review by hexonaut — APPROVED** [review]
Test coverage looks good now. Good to merge imo.

File: src/CharterManager.sol [hunk 2]
```diff
@@ -222,6 +235,35 @@ contract CharterManagerImp {
         ManagedGemJoinLike(gemJoin).exit(urp, usr, amt);
     }
 
+    function move(address u, address dst, uint256 rad) external allowed(u) {
+        address urp = proxy[u];
+        require(urp != address(0), "CharterManager/non-existing-urp");
+
+        VatLike(vat).move(urp, dst, rad);
+    }
```
Comment (gbalabasquer) [inline src/CharterManager.sol:243]: I think there is some weird thing about commits here, could you please merge `master` on this one to see if this block disappears?

File: src/CharterManager.sol [hunk 3]
```diff
@@ -222,6 +235,35 @@ contract CharterManagerImp {
...
+
+    function roll(bytes32 srcIlk, bytes32 dstIlk, address src, address dst, uint256 srcDart) external allowed(src) allowed(dst) {
+        require(gate[srcIlk] == 1 && gate[dstIlk] == 1, "CharterManager/non-gated-ilks");
```
Comment (gbalabasquer) [inline src/CharterManager.sol:246]: Why is this necessary?

File: src/test/CharterManager-unit.t.sol [hunk 4]
```diff
@@ -218,17 +237,17 @@ contract CharterManagerTest is TestBase {
...
-        manager.file(ilk, user, "peace", peace);
-        manager.file(ilk, user, "uline", uline);
+    function init_ilk_gate(bytes32 ilk_, address user, uint256 nib, uint256 peace, uint256 uline) public {
+        manager.file(ilk_, "gate", 1);
+        manager.file(ilk_, user, "nib", nib);
+        manager.file(ilk_, user, "peace", peace);
+        manager.file(ilk_, user, "uline", uline);
     }
 
     function init_user() internal returns (Usr a, Usr b) {
```
Comment (gbalabasquer) [inline src/test/CharterManager-unit.t.sol:253]: Any reason to not remove this one and use the general one everywhere?

File: src/CharterManager.sol (hunk 2 above)
Comment (talbaneth) [reply src/CharterManager.sol]: I saw you merged master, thanks.

File: src/CharterManager.sol (hunk 3 above)
Comment (talbaneth) [reply src/CharterManager.sol:246]: Not strictly speaking, but I wanted to avoid users from trying to move from permissioned vaults (gate=1) to non permissioned ones (gate=0). You could argue that it would fail anyway since the vaults are not marked as rollable, but this makes it more clear.
This also allowed me to pass `1` as the last parameter to _validate() and avoid a "stack too deep" error.

File: src/test/CharterManager-unit.t.sol (hunk 4 above)
Comment (talbaneth) [reply src/test/CharterManager-unit.t.sol:253]: I saw init_user() was used ~60 times so I didn't want to clutter the diff too much.
I guess I can do it as a last thing here once all comments are handled.

File: src/CharterManager.sol [hunk 5]
```diff
@@ -229,6 +242,28 @@ contract CharterManagerImp {
         VatLike(vat).move(urp, dst, rad);
     }
 
+    function roll(bytes32 srcIlk, bytes32 dstIlk, address src, address dst, uint256 srcDart) external allowed(src) allowed(dst) {
```
Comment (gbalabasquer) [inline src/CharterManager.sol:245]: I don't think we need to allow the `src` as basically it is getting its debt repaid. It could be permissionless.

Comment (talbaneth) [reply src/CharterManager.sol:245]: Yeah I know, was trying to be strict here.
The use case is for an institutionval vault owner which control both vaults.
Can change it if we think it's worth it.

Comment (gbalabasquer) [reply src/CharterManager.sol:245]: Yeah fair enough, let's leave it as it is.

File: src/CharterManager.sol (hunk 3 above)
Comment (gbalabasquer) [reply src/CharterManager.sol:246]: Yeah I'm not 100% convinced we should do it, but not opposed to it either. Let's leave it as it is.

File: src/test/CharterManager-unit.t.sol (hunk 4 above)
Comment (gbalabasquer) [reply src/test/CharterManager-unit.t.sol:253]: Ahh ok makes sense.

Comment (gbalabasquer) [reply src/test/CharterManager-unit.t.sol:253]: I'm giving my approval now.

File: src/CharterManager.sol (hunk 2 above)
Comment (gbalabasquer) [reply src/CharterManager.sol]: Yw!

**Review by gbalabasquer — APPROVED** [review]
LGTM.
Feel free to change the `init_user` or not.

**Comment by talbaneth** [comment]
> LGTM. Feel free to change the `init_user` or not.

done.
